qiskit/providers/aer/openpulse/duffing_model_generators.py

```diff
@@ -0,0 +1,438 @@
+                         coupling_symbol='j'):
+    """Returns a PulseSystemModel for a specified duffing oscillator system, and a dict specifying
+    ControlChannel indices for cross-resonance driving.
+
+    Single oscillators are specified by three parameters: frequency v, anharmonicity alpha, and
+    drive strength r, which enter into the Hamiltonian model via the terms:
+        pi*(2*v-alpha)*O + pi*alpha*O*O + 2*pi*r*X*D(t),
+    where O is the number operator, X is the drive operator, and D(t) is the signal of the
+    corresponding DriveChannel.
+
+    Couplings between oscillators are specified in the argument coupling_dict, with keys being
+    edges, and values being the coupling strenghts. Couplings enter the Hamiltonian model via an
+    exchange coupling term:
+        2*pi*j*(A0*C1+C0*A1),
+    where j is the coupling strength, and A0*C1+C0*A1 is the exchange coupling operator between the
+    two oscillators, where A is the annihiliation operator, and C is the creation operator.
+
+    For each pair of coupled oscillators, pulse ControlChannels are prepared for doing cross
+    resonance (CR) drives between oscillators. A CR drive on oscillators 0 with target oscillators
+    1 is modeled with the Hamiltonian term:
+        2*pi*r0*X0*U(t),
+    where r0 is the drive strength for oscillators 0, X0 is the drive operator for oscillators 0,
+    and U(t) is the signal for the CR drive. U(t) is set to have carrier frequency equal to that of
+    the target oscillators.
+
+    Indices for ControlChannels corresponding to CR drives are provided in the returned cr_idx_dict.
+
+    Args:
+        num_oscillators (int): Number of oscillators in the model
+        dim_oscillators (int): Dimension of truncation for each oscillator
+        oscillator_freqs (list): Oscillator frequencies, assumed to be of length num_oscillators
+        anharm_freqs (list): Anharmonicity values, assumed to be of length num_oscillators
+        drive_strengths (list): Drive strength values, assumed to be of length num_oscillators
+        coupling_dict (dict): Specification of the coupling graph with keys being edges, and values
+                              the coupling strengths.
+
+                              For example:
+
+                              * ``{(0,1): 0.02, (1,3): 0.01}`` specifies a system in which
+                                oscillators (0,1) are coupled with strength 0.02, and (1,3) are
+                                coupled with strength 0.01
+        dt (float): Pixel size for pulse instructions
+        freq_symbol (str): Frequency symbol for internal HamiltonianModel
+        anharm_symbol (str): Anharmonicity symbol for internal HamiltonianModel
+        drive_symbol (str): Drive symbol for internal HamiltonianModel
+        coupling_symbol (str): Coupling symbol for internal HamiltonianModel
+
+    Returns:
+        tuple[PulseSystemModel, dict]: The generated oscillator system model, and a dict specifying
```

Comment:
returning the idx as well as the pulse model seems messy. It makes me think that the coupling channel should be a method of the `PulseSystemModel` itself

Reply:
The issue is that U channels are completely arbitrary; in this case we are using them for a particular type of two-qubit interaction, but in general they can be anything. As a result I'm not sure this index information should be put into `PulseSystemModel`, as the meaning of the identifying information (the ordered qubit pair) is strongly tied to the particulars of this model.

The options I can think of are:

1. Put the information into `PulseSystemModel`: I think to maintain generality of this class we would need to add a general dictionary to it for control channel index info, where the type of the keys is arbitrary, so that they can contain identifying info for channel indices of any format, to not make assumptions about the model.
2. Define a new class called `DuffingSystemModel` that inherits from `PulseSystemModel`, but that stores `cr_idx_dict`, and has an added method for looking things up in it. I do kind of like this option as it seems like a natural thing to do when more functionality is needed beyond simply containing info necessary for running the simulation.
3. Leave it as is. I think for now I prefer this option as it's pretty simple, and we can think more about what the `PulseSystemModel` should be.